Refactor k-NN filter search (#3613)
opensearch-project/documentation-website: _search-plugins/knn/filter-search-knn.md

---
layout: default
title: k-NN search with filters
nav_order: 20
parent: k-NN
has_children: false
has_math: true
---

# k-NN search with filters

To refine k-NN results, you can filter a k-NN search using one of the following methods:

- [Scoring script filter](#scoring-script-filter): This approach involves pre-filtering a document set and then running an exact k-NN search on the filtered subset. It does not scale for large filtered subsets.

> - [Boolean filter](#boolean-filter-with-ann-search): This approach runs an [Approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn) search and then applies a filter to the results. Because of post-filtering, it may return significantly less than `k` results for a restrictive filter.

Reviewer: ```suggestion
- [Boolean filter](#boolean-filter-with-ann-search): This approach runs an [approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn) search and then applies a filter to the results. Because of post-filtering, it may return significantly fewer than `k` results for a restrictive filter.
```

> - [Lucene k-NN filter](#using-a-lucene-k-nn-filter): This approach applies filtering _during_ the k-NN search, as opposed to before or after the k-NN search, which ensures that `k` results are returned. You can only use this method with the Hierarchical Navigable Small World (HNSW) algorithm implemented by the Lucene search engine in k-NN plugin versions 2.4 and above.

Reviewer: ```suggestion
- [Lucene k-NN filter](#using-a-lucene-k-nn-filter): This approach applies filtering _during_ the k-NN search, as opposed to before or after the k-NN search, which ensures that `k` results are returned. You can only use this method with the Hierarchical Navigable Small World (HNSW) algorithm implemented by the Lucene search engine in k-NN plugin versions 2.4 and later.
```

## Filtered search optimization

> Depending on your dataset and use case, you might be more interested in maximizing recall or minimizing latency. The following table provides guidance on various k-NN search configurations and the filtering methods to use in order to optimize for higher recall or lower latency. The first three columns of the table provide several example k-NN search configurations. A search configuration consists of:

Reviewer: ```suggestion
Depending on your dataset and use case, you might be more interested in maximizing recall or minimizing latency. The following table provides guidance on various k-NN search configurations and the filtering methods used to optimize for higher recall or lower latency. The first three columns of the table provide several example k-NN search configurations. A search configuration consists of:
```

- The number of documents in an index, where one OpenSearch document corresponds to one k-NN vector.

> - The percentage of documents left in the results after filtering. This value depends on the restrictiveness of the filter that you are providing in the query. The most restrictive filter in the table returns 2.5% of documents in the index, while the least restrictive filter returns 80% of documents.

Reviewer: ```suggestion
- The percentage of documents left in the results after filtering. This value depends on the restrictiveness of the filter that you provide in the query. The most restrictive filter in the table returns 2.5% of documents in the index, while the least restrictive filter returns 80% of documents.
```

- The desired number of returned results (k). 

Once you've estimated the number of documents in your index, the restrictiveness of your filter, and the desired number of nearest neighbors, use the following table to choose a filtering method that optimizes for recall or latency.

| Number of documents in an index | Percentage of documents the filter returns | k | Filtering method to use for higher recall | Filtering method to use for lower latency |
| :-- | :-- | :-- | :-- | :-- |
| 10M | 2.5 | 100 | Scoring script | Scoring script |
| 10M | 38 | 100 | Lucene filter | Boolean filter |
| 10M | 80 | 100 | Scoring script | Lucene filter |
| 1M | 2.5 | 100 | Lucene filter | Scoring script |

> | 1M | 38 | 100 | Lucene filter | Lucene filter/Scoring script |

Reviewer: ```suggestion
| 1M | 38 | 100 | Lucene filter | Lucene filter/scoring script |
```

| 1M | 80 | 100 | Boolean filter | Lucene filter |

## Scoring script filter

> A scoring script filter first filters the documents and then uses a brute-force exact k-NN search on the results. For example, the following query searches for hotels with a rating between 8 and 10, inclusive, that provide parking, and then performs a k-NN search to return the three hotels that are closest to the specified `location`:

Reviewer: ```suggestion
A scoring script filter first filters the documents and then uses a brute-force exact k-NN search on the results. For example, the following query searches for hotels with a rating between 8 and 10, inclusive, that provide parking and then performs a k-NN search to return the 3 hotels that are closest to the specified `location`:
```

> Number of documents in an index | Percentage of documents the filter returns | k | Filtering method to use for higher recall | Filtering method to use for lower latency

Reviewer 2: Jeff has got everyone going on this program: add pipes before and after all table elements (in case we ever switch to a host that can only interpret a markdown table with this format. Not sure if you think that's worthwhile.

          "field": "location",
          "query_value": [
            5.0,
            4.0
          ],
          "space_type": "l2"
        }
      }
    }
  }
}
```
{% include copy-curl.html %}

## Boolean filter with ANN search

A Boolean filter consists of a Boolean query that contains a k-NN query and a filter. For example, the following query searches for hotels that are closest to the specified `location` and then filters the results to return hotels with a rating between 8 and 10, inclusive, that provide parking:

```json
POST /hotels-index/_search
{
  "size": 3,
  "query": {
    "bool": {
      "filter": {
        "bool": {
          "must": [
            {
              "range": {
                "rating": {
                  "gte": 8,
                  "lte": 10
                }
              }
            },
            {
              "term": {
                "parking": "true"
              }
            }
          ]
        }
      },
      "must": [
        {
          "knn": {
            "location": {
              "vector": [
                5,
                4
              ],
              "k": 20
            }
          }
        }
      ]
    }
  }
}
```

> The response contains documents with the matching hotels:

Reviewer: ```suggestion
The response includes documents containing the matching hotels:
```

---

# Search with k-NN filters

To refine k-NN results, you can filter a k-NN search using one of the following methods:

- [Scoring script]({{site.url}}{{site.baseurl}}/search-plugins/knn/knn-score-script/#getting-started-with-the-score-script-for-vectors) on the subset of documents returned by a query: This approach is similar to running an exact k-NN search on the filtered subset but does not scale for large filtered subsets.

- [Boolean filter](#boolean-filter-with-ann-search): This approach runs an [Approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn) search, and then applies the filter to the results. Because it uses post-filtering, a query with a Boolean filter may return significantly less than `k` results for a restrictive filter.

> - [Lucene k-NN filter](#using-a-lucene-k-nn-filter): This approach ensures that `k` results are returned because filtering is applied during the k-NN search as opposed to after the k-NN search, like in post-filtering. The drawback is that you can only use this method with the Hierarchical Navigable Small World (HNSW) algorithm implemented by a Lucene search engine.

Reviewer 3: Please mention that lucene filtering is available from 2.4.

## Filtered search optimization

> Overall, Lucene k-NN filters are more efficient compared to other filtering methods, both in terms of performance and relevancy of search results. 

Reviewer 3: This statement isn't correct for all use cases, and later we provide table where each of 3 filtering types paired with the dataset and filter combination.

> Depending on your dataset and use case, you might be more interested in maximizing recall or minimizing latency. The following table provides guidance on various k-NN search configurations and the filtering methods to use in order to optimize for better recall or lower latency. The first three columns of the table provide several example k-NN search configurations. A search configuration consists of:

Reviewer 3: I think we need to state "higher recall" if it follows by "lower latency".

- The number of documents in an index, where one OpenSearch document corresponds to one k-NN vector.
- The percentage of documents left in the results after filtering. This value depends on the restrictiveness of the filter that you are providing in the query. The most restrictive filter in the table returns 2.5% of documents in the index, while the least restrictive filter returns 80% of documents.
- The desired number of returned results (k). 

Once you've estimated the number of documents in your index, the restrictiveness of your filter, and the desired number of nearest neighbors, use the following table to choose a filtering method that optimizes for recall or latency.

> Number of documents in an index | Percentage of documents the filter returns | k | Filtering method to use for best recall | Filtering method to use for lowest latency

Reviewer 3: same as above, not sure we need to say "best recall", probably "higher recall" is more accurate

}
```

## Lucene k-NN filter implementation

> The OpenSearch k-NN plugin version 2.2 introduced support for running k-NN searches with the Lucene engine using HNSW graphs. Starting with version 2.4, which is based on Lucene version 9.4, you can use Lucene filters for k-NN searches.

Reviewer: ```suggestion
k-NN plugin version 2.2 introduced support for running k-NN searches with the Lucene engine using HNSW graphs. Starting with version 2.4, which is based on Lucene version 9.4, you can use Lucene filters for k-NN searches.
```

When you specify a Lucene filter for a k-NN search, the Lucene algorithm decides whether to perform an exact k-NN search with pre-filtering or an approximate search with modified post-filtering. The algorithm uses the following variables:

- N: The number of documents in the index.
- P: The number of documents in the document subset after the filter is applied (P <= N).
- k: The maximum number of vectors to return in the response.

> The following flow chart summarizes the Lucene algorithm.

Reviewer: ```suggestion
The following flow chart outlines the Lucene algorithm.
```

![Lucene algorithm for filtering]({{site.url}}{{site.baseurl}}/images/lucene-algorithm.png)

> For more information about the Lucene filtering implementation and the underlying `KnnVectorQuery`, see the [Apache Lucene Documentation](https://issues.apache.org/jira/browse/LUCENE-10382).

Reviewer: ```suggestion
For more information about the Lucene filtering implementation and the underlying `KnnVectorQuery`, see the [Apache Lucene documentation](https://issues.apache.org/jira/browse/LUCENE-10382).
```
